### PR TITLE
Extracts import file reading & remapping logic out of CommandLineInterface into their own classes.

### PR DESCRIPTION
**Pull request by christianparpart**
this interface is also used in the LSP PR #8125 (splitout, and now more....):

/cc @cameel (what did you want to work on here?)

- Adds FileReader: extracted from CommandLineInterface, to share the logic between CLI and LSP.
- Adds FileRemapper: extracted from CompilerStack

### CODE REVIEW

File: libsolidity/interface/FileReader.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	{
+		return readFile(_kind, _path);
+	}
+
+private:
+	/// Base path, used for resolving relative paths in imports.
+	boost::filesystem::path m_basePath;
+
+	/// list of allowed directories to read files from
+	std::vector<boost::filesystem::path> m_allowedDirectories;
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]: What do you think about making this a `set`? AFAIK order does not matter here.

File: libsolidity/interface/FileReader.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#include <libsolidity/interface/ReadFile.h>
+
+namespace solidity {
+
+class FileReader
+{
+public:
+	using StringMap = std::map<std::string, std::string>;
+
+	FileReader(
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]:
```suggestion
	explicit FileReader(
```

File: libsolidity/interface/FileReader.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		boost::filesystem::path _basePath,
+		std::vector<boost::filesystem::path> _allowedDirectories
+	):
+		m_basePath(std::move(_basePath)),
+		m_allowedDirectories(std::move(_allowedDirectories)),
+		m_sourceCodes(),
+		m_fullPathMapping()
+	{}
+
+	FileReader(): FileReader({}, {}) {}
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]: Why not add defaults to the other constructor instead?

**Comment by chriseth** [comment]
This PR contains commented-out code and a "TODO" (XXX) comment.

File: libsolidity/interface/FileReader.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	boost::filesystem::path const& basePath() const noexcept { return m_basePath; }
```
Comment (cameel) [inline libsolidity/interface/FileReader.h:54]: I don't see `noexcept` used much around the code base. Should we be using it more?

File: libsolidity/interface/FileReader.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	allowDirectory(_path);
+}
+
+void FileReader::allowDirectory(boost::filesystem::path _path)
+{
+	if (!util::contains(m_allowedDirectories, _path))
+		m_allowedDirectories.emplace_back(std::move(_path));
+}
+
+void FileReader::setSource(string _path, optional<boost::filesystem::path> _fspath, string _source)
```
Comment (cameel) [inline libsolidity/interface/FileReader.cpp]: The `_path` here is not necessarily a path in general case. I think we should have a specific term for it because it gets confusing quickly. In #11105 I proposed calling it "source unit ID" (actually I used "import key" at first but then changed it). Another option could be `VFS key`. What's your opinion on that?

```suggestion
void FileReader::setSource(string _sourceUnitID, optional<boost::filesystem::path> _filesystemPath, string _sourceCode)
```

File: libsolidity/interface/FileReader.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	boost::filesystem::path m_basePath;
+
+	/// list of allowed directories to read files from
+	std::vector<boost::filesystem::path> m_allowedDirectories;
+
+	/// map of input files to source code strings
+	StringMap m_sourceCodes;
+
+	/// map of input file names to full path names suitable for file://-URIs
+	StringMap m_fullPathMapping;
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]: Why not store `boost:filesystem::path` as values?

Comment (cameel) [inline libsolidity/interface/FileReader.h]: Do we really need `m_fullPathMapping`? The concept of "full path" exists only for files that are loaded from the filesystem. There are several important cases where they aren't:
- Source provided in Standard JSON
- Source provided on compiler's standard input
- Source loaded by the callback set by `solc-js`

Moreover the file on disk can change after we have loaded it so it does not seem safe to refer to the file using this path.

File: libsolidity/interface/FileReader.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+{
+	if (_fspath.has_value())
+		m_fullPathMapping[_path] = _fspath.value().generic_string();
```
Comment (cameel) [inline libsolidity/interface/FileReader.cpp]: ` _fspath` is not necessarily the actual path to the file on disk that `readFile()` will read. `readFile()` does some additional transformations like adding base path. If you want a path that can be used to refer to the file outside of this class, you have to do the same transformations here.

File: libsolidity/interface/FileReader.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+	boost::filesystem::path const& basePath() const noexcept { return m_basePath; }
+	void setBasePath(boost::filesystem::path const& _path);
+
+	void allowDirectory(boost::filesystem::path _path);
+
+	std::vector<boost::filesystem::path> const& allowedDirectories() const noexcept { return m_allowedDirectories; }
+
+	StringMap const& sourceCodes() const noexcept { return m_sourceCodes; }
+	StringMap& sourceCodes() noexcept { return m_sourceCodes; }
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]: Why allow messing from the sources from the outside? At first I thought it might be because of `CompilerStack::setSources()` but that method expects `StringMap`, not a mutable reference.

File: libsolidity/interface/FileReader.h (hunk 6 above)
Comment (cameel) [inline libsolidity/interface/FileReader.h]: I'd avoid calling it "path mapping" because it sounds like the path remapping we're doing inside `CompilerStack` (https://docs.soliditylang.org/en/latest/using-the-compiler.html#path-remapping) but apparenly here it's something different.

File: solc/CommandLineInterface.cpp [hunk 9]
```diff
@@ -1297,7 +1245,8 @@ bool CommandLineInterface::processInput()
 				return false;
 			}
 		}
-		StandardCompiler compiler(fileReader);
+		// //XXX StandardCompiler compiler(std::ref(m_fileReader));
+		StandardCompiler compiler([&](auto a, auto b) { return m_fileReader(a, b); });
```
Comment (cameel) [inline solc/CommandLineInterface.cpp]: This looks like something left over from debugging.

File: solc/CommandLineInterface.h [hunk 10]
```diff
@@ -112,16 +113,14 @@ class CommandLineInterface
 
 	bool m_onlyLink = false;
 
+	FileReader m_fileReader;
+
 	/// Compiler arguments variable map
 	boost::program_options::variables_map m_args;
 	/// map of input files to source code strings
-	std::map<std::string, std::string> m_sourceCodes;
+	//std::map<std::string, std::string> m_sourceCodes;
```
Comment (cameel) [inline solc/CommandLineInterface.h]: We should just remove it from here.

File: solc/CommandLineInterface.cpp [hunk 11]
```diff
@@ -622,14 +623,14 @@ bool CommandLineInterface::readInputFilesAndConfigureRemappings()
 				}
 
 				// NOTE: we ignore the FileNotFound exception as we manually check above
-				m_sourceCodes[infile.generic_string()] = readFileAsString(infile.string());
+				m_fileReader.setSource(infile.generic_string(), infile, readFileAsString(infile.string()));
```
Comment (cameel) [inline solc/CommandLineInterface.cpp]: I think it would be better to pass in only `infile` and let the reader do other transformations, like calling `generic_string()`.

File: solc/CommandLineInterface.cpp [hunk 12]
```diff
@@ -736,12 +737,12 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
 map<string, Json::Value> CommandLineInterface::parseAstFromInput()
 {
 	map<string, Json::Value> sourceJsons;
-	map<string, string> tmpSources;
+	map<string, string> tmpSource;
```
Comment (cameel) [inline solc/CommandLineInterface.cpp]: Why the change to singular? It does store multiple sources.

File: solc/CommandLineInterface.cpp [hunk 13]
```diff
@@ -752,11 +753,12 @@ map<string, Json::Value> CommandLineInterface::parseAstFromInput()
 			astAssert(ast["sources"][src][astKey]["nodeType"].asString() == "SourceUnit",  "Top-level node should be a 'SourceUnit'");
 			astAssert(sourceJsons.count(src) == 0, "All sources must have unique names");
 			sourceJsons.emplace(src, move(ast["sources"][src][astKey]));
-			tmpSources[src] = util::jsonCompactPrint(ast);
+			tmpSource[src] = util::jsonCompactPrint(ast); // could this potentially override pre-existing sources?
```
Comment (cameel) [inline solc/CommandLineInterface.cpp]: Judging by the function name (`parseAstFromInput()`), I guess it could?

Interesting, I did not know we had a feature like this. Maybe I should mention it in #11105?

**Comment by cameel** [comment]
> /cc @cameel (what did you want to work on here?)

I'm refactoring the path transformation logic and moving it into a single place to be able to easily add unit tests and fix bugs in it. Lately I've been actually spending more time just trying to understand the whole path logic, which is suprisingly complex given how little code it contains. I have documented my current understanding in #11105.

Some of changes overlap with yours (moving the logic out of `CommandLineInterface`), though I did not extract the reader itself. I created `PathResolver` responsible for path mapping, base path, allowed paths, etc. and moved logic from both `CommandLineInterface` and `CompilerStack` there. My current plan is to drop the part that overlaps with your changes and just build on top of what you're doing here once we get your PR merged.

File: libsolidity/interface/FileReader.h (hunk 4 above)
Comment (axic) [reply libsolidity/interface/FileReader.h:54]: Probably because the codebase uses exceptions heavily :)

File: libsolidity/interface/FileReader.h [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+#include <boost/filesystem.hpp>
+#include <map>
+#include <vector>
+
+#include <libsolidity/interface/ReadFile.h>
+
+namespace solidity {
+
+class FileReader
```
Comment (axic) [inline libsolidity/interface/FileReader.h:35]: I think this class is not well documented, but should be. It has many methods, which should be documented, but more importantly a generic description explaining the goal os behaviour should be here at the top.

Comment (axic) [reply libsolidity/interface/FileReader.h:35]: Also I would argue it would be better to rethink how this class works once we understand it, so we can trim it and not just move the potentially badly designed legacy code from the cli. As @cameel mentioned here: https://github.com/ethereum/solidity/pull/11113/files#r595374456.

File: libsolidity/interface/FileReader.h (hunk 4 above)
Comment (axic) [inline libsolidity/interface/FileReader.h:54]: Do we actually need to use `boost::filesystem::path` in the public API or can we just use strings?

File: libsolidity/interface/FileReader.h [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+	/// Adds given source, imported as @p _path with filesystem path @p _fspath.
+	void setSource(std::string _path, std::optional<boost::filesystem::path> _fspath, std::string _source);
+
+	/// Resets all sources to the given ones.
+	void setSources(StringMap _sources);
+
+	frontend::ReadCallback::Result readFile(std::string const& _kind, std::string const& _path);
+
+	frontend::ReadCallback::Result operator()(std::string const& _kind, std::string const& _path)
```
Comment (axic) [inline libsolidity/interface/FileReader.h]: Why do we need both the operator and `readFile`?

**Comment by axic** [comment]
Without wanting to complicate this PR too much, I think the problems raised and trying to be solved in #9353 should be addressed here. And I think by trying to document what exactly this class is doing, would lead to a better understanding and design, ultimately fixing the problems identified by #9353 :wink:

File: libsolidity/interface/FileReader.h (hunk 1 above)
Comment (christianparpart) [reply libsolidity/interface/FileReader.h]: I can do that, however, I was just moving this code, because I needed the very same functionality in the LSP and I wanted to avoid code duplication. Hmm...

File: libsolidity/interface/FileReader.h (hunk 4 above)
Comment (christianparpart) [reply libsolidity/interface/FileReader.h:54]: Also `noexcept` is fairly in this way. And since we're returning just a cref it is safe to explictly state that. It is common practice for me to make noexcept what is meant to be noexcept. But I'm fine in not doing that? :)

File: libsolidity/interface/FileReader.h (hunk 6 above)
Comment (christianparpart) [reply libsolidity/interface/FileReader.h]: Hm, I need that mapping from (let's call it) source unit ID, so I can properly map forth and back in the language server.
I am not sure it is actually good to use `filesystem::path` here when the underlying source code is not necessarily located on your local machine. I mean, instanciating a `filesystem::path` still works, but I am not sure it's best choice, what do you think?

Comment (christianparpart) [reply libsolidity/interface/FileReader.h]: > Do we really need `m_fullPathMapping`? The concept of "full path" exists only for files that are loaded from the filesystem. 

When compiling code you usually only deal with the so called source unit IDs, and there just having the basic FileReader interface that already did exist was fine, as it was mapping that to the local disk if needed to fetch the source code.

The language client however passes file URIs to the language server, and I need to be able map these back to the source unit ID (and vice versa). But maybe there is a better way?

[...]
> Moreover the file on disk can change after we have loaded it so it does not seem safe to refer to the file using this path.

I think this could have happened before either, and we didn't care about because `solc` usually is short-living. However, when running in language server mode, `solc` is a long running process and then it might make sense to consider this.
However, the client is telling us about any changes the user did (within the client IDE, - not from the file system).

File: libsolidity/interface/FileReader.h (hunk 8 above)
Comment (christianparpart) [reply libsolidity/interface/FileReader.h]: I am pretty sure I used that to move the map out of this, as FileReader() wasn't used after that anymore. But the code has changed, so I'll remove this.

Comment (christianparpart) [reply libsolidity/interface/FileReader.h]: Oh, It's CommandLineInterface.cpp's link() function that does actually still mutate `m_sourceCodes` from the outside.

It would be worked around by just retrieving the copy, modifying the copy, and copying then back. But that sounds expensive. Any ideas?

File: libsolidity/interface/FileReader.h (hunk 6 above)
Comment (christianparpart) [reply libsolidity/interface/FileReader.h]: It's not the "remapping", exactly. But it's mapping source unit ID to a filesystem path representation (that does not necessarily has to exist on the local file system, but still is a file system path that is being mapped to). Do you have any name suggestions here? :)

File: libsolidity/interface/FileReader.h (hunk 1 above)
Comment (cameel) [reply libsolidity/interface/FileReader.h]: Sure. To be honest I also did not do that in my refactor (yet). If you want to keep it as a pure refactor without too many changes that's fine with me.

File: libsolidity/interface/FileReader.h (hunk 2 above)
Comment (cameel) [reply libsolidity/interface/FileReader.h]: After discussing in on the chat, it's not that important.

File: libsolidity/interface/FileReader.h (hunk 4 above)
Comment (cameel) [reply libsolidity/interface/FileReader.h:54]: I'm fine with that, just wondering if I should start doing that too. So far I just had the impression that we don't care about it too much but if we do, I'll start using it.

File: libsolidity/interface/FileReader.h (hunk 6 above)
Comment (cameel) [reply libsolidity/interface/FileReader.h]: By "not necessarily located on your local machine" do you mean that it's actually loaded over network or just e.g. stored in JSON?

Comment (cameel) [reply libsolidity/interface/FileReader.h]: > I think this could have happened before either, and we didn't care about because solc usually is short-living. 

AFAIK it could not. We load a file into the virtual filesystem once and then `CompilerStack` uses that whenever the file is referenced. If you manage to make it load the file again it's only because it thinks it's a different file (i.e. it has a different source unit ID) and it will be compiled again and treated as a separate source unit.

> The language client however passes file URIs to the language server, and I need to be able map these back to the source unit ID (and vice versa). But maybe there is a better way?

How does the language server access files from Standard JSON or compiler's standard input? Or from JSON passed to the standard input? :)

Also, do you support files loaded from disk like this?
```json
       {
           "language": "Solidity",
           "sources": {
               "/usr/local/dapp-bin/token.sol": {
                   "urls": ["/projects/mytoken.sol"]
               }
           },
           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
       }
```

I think it would be best if there was a way for the compiler to share the source it has already loaded into its virtual filesystem.

If that is not possible and the LSP is meant to work only with files that were loaded from disk then I think you could just replace this map with a function that calculates the path based on source unit ID or the other way. That should be deterministic as long as you know that it was not loaded from some other source (we might have to add some enum to indicate where it came from).

File: libsolidity/interface/FileReader.h (hunk 8 above)
Comment (cameel) [reply libsolidity/interface/FileReader.h]: Maybe you could move it out, link it and then move it back in?

```c++
StringMap relinquishSourceCodes() noexcept {
    StringMap tmpMap = move(m_sourceCodes);
    m_sourceCodes = {};
    return tmpMap;
}
```

The downside is that you would have to either keep other members in sync or remember that the reader is in an unusable state in the meantime.

Another option would be to just built the source code map first, link it and only then give it to the file reader.

File: libsolidity/interface/FileReader.h (hunk 6 above)
Comment (christianparpart) [reply libsolidity/interface/FileReader.h]: in the context of the language server, the communication happens (usually) over stdio, so the file contents may come over stdin (in form of a json rpc call). EDIT: I've changed it to use filesystem::path nevertheless.

File: libsolidity/interface/FileReader.h [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <boost/filesystem.hpp>
+
+#include <map>
+#include <set>
+
+#include <libsolidity/interface/ImportRemapper.h>
+#include <libsolidity/interface/ReadFile.h>
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]: We usually use a different order for imports: our own headers go first, then third-party libraries like boost, then stdlib.

File: libsolidity/interface/FileReader.h [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+namespace solidity::frontend
+{
+
+/// FileReader - used for progressively loading source source code.
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]:
```suggestion
/// FileReader - used for progressively loading source code.
```

File: libsolidity/interface/FileReader.h (hunk 14 above)
Comment (cameel) [reply libsolidity/interface/FileReader.h:35]: I think a light refactor that only moves stuff out of `CommandLineInterface` is fine. This PR was only meant to publish the changes needed for the LSP to reuse the file reader. I'm going to refactor it further after that anyway.

File: libsolidity/interface/FileReader.h [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+		FileSystemPathSet _allowedDirectories = {}
+	):
+		m_basePath(std::move(_basePath)),
+		m_allowedDirectories(std::move(_allowedDirectories)),
+		m_sourceCodes(),
+		m_sourceCodeOrigin()
+	{}
+
+	void setBasePath(boost::filesystem::path _path) { m_basePath = std::move(_path); }
+	std::string basePath() const noexcept { return m_basePath.generic_string(); }
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]: I think this should be symmetrical. It either both accepts and returns `std::string` or uses `boost::filesystem::path` for both.

```suggestion
	boost::filesystem::path basePath() const noexcept { return m_basePath; }
```

File: libsolidity/interface/FileReader.h [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+	/// Retrieves the source code for a given source unit ID.
+	SourceCode const& sourceCode(SourceUnitID const& _sourceUnitID) const { return m_sourceCodes.at(_sourceUnitID); }
+
+	/// Resets all sources to the given map of source unit ID to source codes.
+	void setSources(StringMap _sources);
+
+	/// Adds the source code for a given file system path.
+	///
+	/// The source unit ID will be equivalent to the source unit ID.
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]: x will be equivalent to x?

File: solc/CommandLineInterface.cpp [hunk 20]
```diff
@@ -738,10 +745,10 @@ map<string, Json::Value> CommandLineInterface::parseAstFromInput()
 	map<string, Json::Value> sourceJsons;
 	map<string, string> tmpSources;
 
-	for (auto const& srcPair: m_sourceCodes)
+	for (auto const& sourceUnitID: m_fileReader.sourceCodes() | ranges::views::keys)
```
Comment (cameel) [inline solc/CommandLineInterface.cpp]: I originally suggested `views::keys` but now that you changed it I see that you could make it even simpler by iterating over values:

```suggestion
	for (auto const& sourceCode: m_fileReader.sourceCodes() | ranges::views::values)
```

File: libsolidity/interface/FileReader.h [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+	boost::filesystem::path m_basePath;
+
+	/// list of allowed directories to read files from
+	FileSystemPathSet m_allowedDirectories;
+
+	/// map of input files to source code strings
+	StringMap m_sourceCodes;
+
+	/// map of input file names to full path names suitable for file://-URIs
+	PathMap m_sourceCodeOrigin;
```
Comment (cameel) [inline libsolidity/interface/FileReader.h]: I'm still not convinced that this is the right way to tell LSP where the files reside. It does not take base path into account, not all overloads of `setSource()` fill it and `setSources()` can just casually clear it. It does not provide access to files not loaded by `FileReader`.

You suggested that you could move it to the LSP PR. I said it's not necessary but now I think it's actually a good idea because for me it's the only thing holding back this PR. I'm not sure if I'll be reviewing the LSP PR so how about putting it in a PR of its own? We could then discuss it in detail. I think this functionality should really sit in `CompilerStack`, not here.

**Review by cameel — CHANGES_REQUESTED** [review]
The only important issue left in my opinion is `m_sourceCodeOrigin` and it could be moved to a separate PR. The rest are just cosmetic tweaks.

File: libsolidity/interface/FileReader.h (hunk 21 above)
Comment (christianparpart) [reply libsolidity/interface/FileReader.h]: ok. removed.

**Review by cameel — DISMISSED** [review]

**Review by cameel — APPROVED** [review]

File: libsolidity/interface/ImportRemapper.h [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+#include <optional>
+#include <string>
+#include <vector>
+
+namespace solidity::frontend
+{
+
+// Some helper typedefs to make reading the signatures more self explaining.
+using SourceUnitName = std::string;
```
Comment (chriseth) [inline libsolidity/interface/ImportRemapper.h:28]: Not very happy about those. The second one is not even used.

Comment (christianparpart) [reply libsolidity/interface/ImportRemapper.h:28]: This was an outcome of the discussion we had via voice. @cameel you want to build on top of this PR, right? Can you then also delete the line below this one in your PR? (`SourceCode typedef` - the code that used it was removed.)

Comment (christianparpart) [reply libsolidity/interface/ImportRemapper.h:28]: The reasoning behind was that we deal a lot with `std::string` but meaning something different. My proposal first was one *could* wrap that into `struct WrappedType { std::string value; };` but I feared that would have been too much. OTOH I don't mind, we can move back to plain std::string in @cameel's PR then?

Comment (cameel) [reply libsolidity/interface/ImportRemapper.h:28]: >  Can you then also delete the line below this one in your PR?

Yeah, I can change that.

> OTOH I don't mind, we can move back to plain std::string in @cameel's PR then?

I liked the idea of having distinct types for those but now I'm not so sure about it. There are a bit too many of them and they still do not cover everything (e.g. the remappings still contain plain strings because they're not paths, not source unit names, just... fragments). So I'm fine with it as is but I'm also fine with going back to strings. We can sort this out in the refactor PR.

File: libsolidity/interface/ImportRemapper.cpp [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+	return path;
+}
+
+optional<ImportRemapper::Remapping> ImportRemapper::parseRemapping(string const& _remapping)
+{
+	auto eq = find(_remapping.begin(), _remapping.end(), '=');
+	if (eq == _remapping.end())
+		return {};
+
+	auto colon = find(_remapping.begin(), eq, ':');
```
Comment (mijovic) [inline libsolidity/interface/ImportRemapper.cpp:86]: Should there be an assertion that colon is found, or which case it can't be not found?

**Comment by mijovic** [comment]
Are there any tests in codebase that are testing this?
